RecoMET/METAlgorithms/src/CSCHaloAlgo.cc

```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
+    clustercand.SetBeamHaloRecHitsCandidates(bhrhcandidates);
+
+    TheHaloClusterCandsHB.push_back(clustercand);
+  }
+
+  return  TheHaloClusterCandsHB;
+} 
+
+
+std::vector<HaloClusterCandidateHE> CSCHaloAlgo::GetHaloClusterCandidateHE(edm::Handle<EcalRecHitCollection>& ecalrechitcoll, edm::Handle<HBHERecHitCollection>& hbherechitcoll,float et_thresh_seedrh){
+
+  std::vector<HaloClusterCandidateHE> TheHaloClusterCandsHE;
+
+  reco::Vertex::Point vtx(0,0,0);
+
+  for(size_t ihit = 0; ihit<hbherechitcoll->size(); ++ ihit){
+    HaloClusterCandidateHE  clustercand;
+    
+    const HBHERecHit & rechit = (*hbherechitcoll)[ ihit ];
+    math::XYZPoint rhpos = getPosition(rechit.id(),vtx);
+    //Et condition
+    double rhet = rechit.energy()/cosh(rhpos.eta());
+    if(rhet<et_thresh_seedrh) continue;
+    if(abs(rhpos.z())<380) continue;
+    double eta = rhpos.eta();
+    double phi = rhpos.phi();
+    double rhr = sqrt(rhpos.x()*rhpos.x()+rhpos.y()*rhpos.y());
+    bool isiso = true;
+    double etcluster(0),hdepth1(0);
+    int clustersize(0);
+    double etstrip_phiseedplus1(0), etstrip_phiseedminus1(0);
+
+    //Building the cluster 
+    edm::RefVector<HBHERecHitCollection> bhrhcandidates;
+    for(size_t jhit = 0; jhit<hbherechitcoll->size(); ++ jhit){
+      const HBHERecHit & rechitj = (*hbherechitcoll)[ jhit ];
+      HBHERecHitRef rhRef(hbherechitcoll,jhit);
+      math::XYZPoint rhposj = getPosition(rechitj.id(),vtx);
+      double rhetj = rechitj.energy()/cosh(rhposj.eta());
+      if(rhetj<2) continue;
+      if(abs(rhposj.z())<380) continue;
+      if(rhpos.z()*rhposj.z()<0) continue;
+      double phij = rhposj.phi();
+      if(abs(deltaPhi(phi,phij))>0.4) continue;
+      double rhrj = sqrt(rhposj.x()*rhposj.x()+rhposj.y()*rhposj.y());
+      if(abs( rhr-rhrj )>50) continue;
+      if(abs(deltaPhi(phi,phij))>0.2 ||abs( rhr-rhrj )>20 ){isiso=false;break;}//The deposit should be isolated
+      if(deltaPhi(phi,phij)>0.05) etstrip_phiseedplus1+=rhetj;
+      if(deltaPhi(phi,phij)<-0.05) etstrip_phiseedminus1+=rhetj;
+      clustersize++;
+      etcluster+=rhetj;
+      if(abs( rhposj.z())<405 )hdepth1+=rhetj;
+      //No timing condition for now in HE
+      bhrhcandidates.push_back(rhRef);
+    }
+    //Isolation conditions
+    if(!isiso) continue;
+    if(etstrip_phiseedplus1/etcluster>0.1&& etstrip_phiseedminus1/etcluster>0.1) continue;
+    
+    //Calculate E/H
+    double eoh(0);
+    for(size_t jhit = 0; jhit<ecalrechitcoll->size(); ++ jhit){
+      const EcalRecHit & rechitj = (*ecalrechitcoll)[ jhit ];
+      math::XYZPoint rhposj = getPosition(rechitj.id(),vtx);
+      double rhetj = rechitj.energy()/cosh(rhposj.eta());
+      if(rhetj<2) continue;
+      if(rhpos.z()*rhposj.z()<0) continue;
+      double etaj = rhposj.eta();
+      double phij = rhposj.phi();
+      double dr = sqrt(abs(eta-etaj)*abs(eta-etaj)+deltaPhi(phi,phij)*deltaPhi(phi,phij));
+      if(dr>0.3) continue;
```

Comment:
there is no need to add sqrt here.
Also abs is not needed in abs*abs expression.